<!DOCTYPE html>
<html>
 <head>
 <title>User's Guide - Amplicon utilities</title>
 <link rel="stylesheet" type="text/css" href="style/docs_style.css">
 <style type="text/css">
 .row { vertical-align: top; height:auto !important; }
 .list {display:none; }
 .show {display: none; }
 .hide:target + .show {display: inline; }
 .hide:target {display: none; }
 .hide:target ~ .list {display:inline; }
 @media print { .hide, .show { display: none; } }
 </style>
<h2>
 <a href="http://people.oregonstate.edu/~meyere/tools.html" title="Back to Tools page"
  style="text-decoration: none; color: black; margin-left: 2%;">Tools | </a>
 <a href="#top" title="Collapse all menus" style="text-decoration: none; color: black;">Analysis of amplicon sequence variants with ASV utilities (v1.0)</a>
 </h2>
 <hr align="left" style="margin-left: 2%;">
 </head>

 <body>
   <div class="row" style="font-size: 90%;">
  <a href="#about" class="hide" id="about" style="text-decoration: none; color: black; margin-left: 2%">
   <span title="Click to expand">
   About this guide</span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%">
   <span title="Click to collapse">
   About this guide</span></a>
   <div class="list">
   <p align="left" style="margin-left: 2%">This page outlines our standard pipeline for analysis of amplicon sequence variation. 
    The guide focuses on analysis of Symbiodinium communities in reef-building corals using ITS2 amplicons, but the procedures 
    described here can be easily modified for other targets or sample types. This guide covers the most recently updated version of
    the ASV utilities pipeline as of July 2018 (v1.0).<br><br>
   These tools are intended to be used on a high-performance computing cluster, and assume a basic knowledge of Linux and command-line
   tools. To analyze large numbers of samples in parallel, the user will probably want to (a) combine some of the
   following steps into shell scripts, and (b) submit each job to a job scheduler such as SGE. Since the details of
   those steps may vary from one cluster to another, we have left those details up to the end user.<br><br>
   Click on each header to expand or collapse that section.
   </p>
   </div>
  <br>
  </div>
  <div class="row">
  <a href="#install" class="hide" id="install" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <b><span title="Click to expand">
   &#9655 Download scripts and install required software</span></b></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <b><span title="Click to collapse">
   &#9665 Download scripts and install required software</span></b></a>
   <div class="list">
   <ul>
   <li><b>BioPerl</b></li>
    <p>Our scripts rely on BioPerl modules. First, check whether you need to install BioPerl. <br>
    <code>
    perl -MBio::SeqIO -e 0 <br>
    </code>
    If you get no feedback, BioPerl is available on your system. If you get an error message something like the following:<br>
    <code>
    Can't locate Bio/SeqIO.pm in @INC (@INC contains: /etc/perl /usr/local/lib/perl/5.14.2 ...    <br>
    </code>
    this indicates that you need to install BioPerl. <br><br>
    To install BioPerl, go to <a href="http://www.bioperl.org/wiki/Getting_BioPerl" title="Link to BioPerl website">the BioPerl Wiki</a> and follow the instructions given there. <br>
    </p>
   <li><b>Scripts</b></li>
   <p>Our scripts for Amplicon Sequence Variant (ASV) analysis are hosted on GitHub. You'll want to download scripts from this repository:
   <ul>
    <li><a href="https://github.com/Eli-Meyer/Amplicon_utilities/" title="Link to GitHub repository">Scripts for Amplicon Sequence Variant analysis.<a></li>
   </ul>
   <br>The best way to do this is to use the <i>git</i> tool. While this is undoubtedly the best and most widely-used platform<br>
   for sharing code, the git tool itself is not simple or intuitive. To learn about use of git, we recommend
   <a href="http://gitref.org/" title="Link to GitRef website">this resource</a>.<br>
   To download the above repositories onto your system, assuming <i>git</i> is available:<br>
   <code>git clone git://github.com/Eli-Meyer/Amplicon_utilities.git<br></code>
   Make sure you have execute permission to these scripts, and that they are in your path ($PATH).<br>
   </p>
   <li><b>R</b></li>
   You will need R installed both on your remote computing cluster (Linux) and your local computer (typically Mac OS or Windows).  
   R is widely used and extensively documented. To evaluate whether R is installed on your system try these steps:<br>
    <code>which R</code><br>
    <code>which Rscript</code><br>
   If you get feedback something like the following it indicates R is available on your system (this is expected). <br>
    <code>/bin/Rscript</code><br>
   If you get no feedback, or "command not found" this indicates R is not installed. For the computing cluster, contact your
   system administrator. For your local computer, see installation instructions at:<br>
    <a href="https://www.r-project.org/">The R Project for Statistical Computing</a><br>   

       <li><b>R packages</b></li>
   This pipeline requires several R packages. The following links contain documentation for each package, but aren't required for 
   installation. 
    <ul>
     <li><a href="https://bioconductor.org/packages/release/bioc/html/dada2.html">dada2</a></li>
     <li><a href="https://cran.r-project.org/web/packages/vegan/index.html">vegan</a></li>
     <li><a href="https://cran.r-project.org/web/packages/ape/index.html">ape</a></li>
     <li><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></li>
     <li><a href="https://cran.r-project.org/web/packages/gplots/index.html">gplots</a></li>
     <li><a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">RColorBrewer</a></li>
    </ul>
    To evaluate whether a package is available in R on your system, open R and type:<br>
    <code>library(<i>package_name</i>)</code><br>
    For most R packages, you can install the package as:<br>
    <code>install.package("<i>package_name</i>)"</code><br>
    If the package isn't available for that method try:<br>
    <code>source("https://bioconductor.org/biocLite.R")</code><br>
    <code>biocLite("<i>package_name</i>")</code><br>
    </ul>
   </div>
  </div>

  <div class="row"><br>
  <a href="#process" class="hide" id="process" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Process reads</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Process reads</b></span></a>
   <div class="list">
   <ul>
   <li><b>Trim to remove low-quality ends</b></li>
    First, take a subset of forward and reverse reads to evaluate the distribution of sequence quality scores:<br>
    <code>RandomFastqPE.pl -f name_R1.fastq -r name_R2.fastq -n 1000 -o sub.R1.fastq -p sub.R2.fastq</code><br>
    View the average quality score by position:<br>
    <code>QualViewFastq.pl -i subset_R1.fastq > R1_qplot.txt</code><br>
    <code>QualViewFastq.pl -i subset_R2.fastq > R2_qplot.txt</code><br>
    Look at the distribution of quality scores and decide where to cut sequences. Ideally you should remove regions 
    where median quality scores are < 30, but balance this with the need to preserving adequate overlap (≥30 bp) between 
    F and R reads (based on amplicon size).<br>
    <code>less R1_qplot.txt</code><br>
    <code>less R2_qplot.txt</code><br>
    Having decided on settings, you can now run the quality trimming script as e.g.:<br>
    <code>TruncateFastq.pl -i sub.R1.fastq -s 1 -e 250 -o trunc.R1.fastq</code><br>
    <code>TruncateFastq.pl -i sub.R2.fastq -s 1 -e 200 -o trunc.R2.fastq</code><br>
   <br>
   <li><b>Remove any low-quality reads that remain</b></li>
    To minimize the effects of sequencing error on our analysis of sequence variation, we remove any reads that have too
    many low-quality positions in the regions that remained after quality trimming. We’ll reject any reads in which > X positions
    had quality scores below a threshold M. e.g.<br>
    <code>QualFilterFastq.pl -i trunc_R1.fastq -m 20 -x 25 -o hq_R1.fastq</code><br>
    <code>QualFilterFastq.pl -i trunc_R2.fastq -m 20 -x 20 -o hq_R2.fastq</code><br>
    <br>
   <li><b>Remove reads lacking expected primer sequence</b></li>
   Valid amplicons will begin and end with known sequences defined by the primers used to amplify those targets. To exclude reads 
   lacking these sequences, we can enter the following, using DNA sequences of the primers used to amplify the targets. <br>
    <code>ScreenFastqByPrimer.pl -i hq_R1.fastq -s <i>sequence1</i> -m 2 </code><br>
    <code>ScreenFastqByPrimer.pl -i hq_R2.fastq -s <i>sequence2</i> -m 2 </code><br>
    <br>
   <li><b>Remove orphan (unpaired) reads</b></li>
   These occur when one read was removed during filtering, but its mate was not. Since paired reads are required for 
   ASV analysis, we exclude orphan reads as:<br>
   <code>FindPairedReads.pl -f target.R1.fastq -r target.R2.fastq -g name_R1.final.fastq -i name_R2.final.fastq</code><br>
   <br>
   </ul>
   </div>
  </div>

  <div class="row"><br>
  <a href="#asv" class="hide" id="asv" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Analyze amplicon sequence variants (ASVs)</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Analyze amplicon sequence variants (ASVs)</b></span></a>
   <div class="list">
   <ul>
   <li><b> Prepare working directory </b></li>
   Make sure all your cleaned reads are in a single directory and that they are named as:<br>
    <code> <i>name1</i>_R1.fastq </code><br>
    <code> <i>name1</i>_R2.fastq </code><br>
   <li><b> Prepare the R script</b></li>
   Edit the R script "ASV_dada2.R" as appropriate for your input files. Note the following sections:<br>
    <ul>
     <li>“Set working directory and build file lists”</li>
     The script expects that your files are named as “name1_R1.fastq”, etc. If you use a different naming convention 
     adjust this section accordingly.<br>
     <li>“Filtering”</li>
     If you used different primers than our usual ITS2 primers (19 and 27 bp), adjust the primer lengths accordingly.<br>
     <li>“Identify amplicon sequence variants (ASVs)”</li>
     If you are analyzing ITS data, use BAND_SIZE=32, as recommended by the software authors for targets like ITS with 
     high indel rates. If analyzing 16S data, leave at the default 16 (comment out the setDadaOpt line or set at 16).<br>
     <li>“Inspect and trim distribution of sequence lengths”</li>
     Some variation in ASV lengths is likely to be biologically real, but you may wish to exclude extreme outliers. You
     can define these limits a priori based on your expected amplicon size, and later refine them if desired based on the 
     distribution of ASV lengths observed in an initial analysis. <br>
    </ul>
    <li><b>Run the R script</b></li>
      After editing the R script as appropriate for your data, you can run the script as:<br>
    <code>./ASV_dada2.R >ASVs.log</code><br>
    <li><b> Evaluate output </b></li>
      The two main outputs from the pipeline are:<br>
    <ul>
     <li>ASV.fasta: a FASTA file containing DNA sequence of each amplicon sequence variant identified in the data</li>
     <li>ASVtable.txt: a file containing the number of times each ASV occured in each sample</li>
    </ul>
      Additional details about the analysis are reported in the log file (ASVs.log, in the example above).<br><br>
      The script also produces several figures (PDF format) describing the quality of input data and the distribution of 
      ASVs across samples. <br>
    <ul>
     <li>QualPlotN.pdf: Shows quality scores by position in the processed reads for the first four samples.</li>
     <li>ErrorEstimatesX.pdf: Comparison of error rate model estimates, observed error rates, and theoretical expectations.</li>
     <li>ASVlengths.pdf: The distribution of ASV lengths.</li>
     <li>ASVsBySample.pdf: A simple overview of communities in each sample inferred from ASV abundances.</li>
    </ul>
   </ul>
   </div>
  </div>
 
  
  <div class="row"><br>
  <a href="#annot" class="hide" id="annot" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Identify origin of ASVs</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Identify origin of ASVs</b></span></a>
   <div class="list">
    <br>
    <p style="margin-left: 2%"> The instructions in this step focus specifically on analysis of Symbiodinium communities in reef-building corals. The database 
    will have to be prepared differently if your target, samples, or biological questions are different, although the same general
     process is generally applicable for a wide range of target and sample types.</p>
   <ul>
   
   <li><b> Prepare BLAST database </b></li>
   First prepare a FASTA file containing:<br>
    <ol>
     <li>ITS2 sequences from a diverse collection of Symbiodinium types, with sequences named in an informative manner.
      For example, <i>clade</i>_<i>sequenceID</i>.</li>
     <li>ITS2 sequences from the host coral or a close relative. 
    </ol>
    Next, prepare a BLAST database from that file as:<br>
    <code>makeblastdb -in customDB.fasta -dbtype nucl</code><br>
    <br>
    
    <li><b> Annotate ASVs based on comparison with the BLAST db </b></li>
    Finally, you can compare your ASVs with the custom BLASTN database and add this information to the table of ASV abundance. 
    This identifies the biological origin of each sequence, providing context for interpretation of the ASV profiles.<br>
    <code>BlastnID.pl -q ASVs.fasta -d customDB.fasta -s 200 -i ASVtable.txt -o annot.ASVtable.txt </code><br>
    The output from this script is a tab-delimited text file with information about each ASV. This is identical to the input
    file, but with five additional columns at the end (far right):
    <ul>
     <li>BestMatch: ID of the best match</li>
     <li>BestScore: Bit-score of the best match</li>
     <li>EvalRatio: Ratio of e-value for best match to the evalue of second best match. (1 indicates a tie, closer to 0 
       indicates an unambiguous best match. A ratio of 0.02 indicates the best match was 50 times less likely to occur 
       by chance than the second best match). </li>
     <li>NextHit: ID of the second best match</li>
     <li>NextScore: Bit-score of the second best match</li>
    </ul>
    These allow the user to evaluate the confidence of each assignment by comparing the best and second-best matches,
    and may be useful for excluding some ASVs (e.g. ASVs derived from the coral host instead of algal symbionts).<br>
    </ul>
   </div>
  </div>
    
  
 <hr align="left" style="margin-left: 2%;">
  <p style="margin-left: 2%; font-size: 85%;">Last updated 14 July 2018, E. Meyer.</p>

 </body>
</html>
